framework/src/node/forger/forger.ts

```diff
@@ -191,16 +198,52 @@ export class Forger {
 			);
 		}
 
-		const account = await this._chainModule.dataAccess.getAccountByAddress(forgerAddress);
-
-		if (forging) {
-			this._keypairs.set(forgerAddress, keypair);
-			this._logger.info(`Forging enabled on account: ${account.address.toString('hex')}`);
-		} else {
+		if (!forging) {
+			// Disable delegate by removing keypairs corresponding to address
 			this._keypairs.delete(forgerAddress);
-			this._logger.info(`Forging disabled on account: ${account.address.toString('hex')}`);
+			this._logger.info(`Forging disabled on account: ${forgerAddress.toString('hex')}`);
+			return {
+				address: forgerAddress,
+				forging,
+			};
 		}
 
+		const lastBlockHeader = this._chainModule.lastBlock.header;
+		const previouslyForgedMap = await getPreviouslyForgedMap(this._db);
+		const forgingInput = { height, maxHeightPreviouslyForged, maxHeightPrevoted };
+
+		if (!isSyncedWithNetwork(lastBlockHeader, forgingInput)) {
+			throw new Error('Failed to enable forging as the node is not synced to the network.');
+		}
+
+		if (!overwrite) {
+			// check if forger info exists
+			if (!previouslyForgedMap.has(forgerAddress)) {
+				throw new Error('Failed to enable forging due to missing forger info.');
+			}
+			// check if forger info matches input
+			const forgerInfo = previouslyForgedMap.get(forgerAddress);
+			if (
+				forgerInfo?.height !== height ||
+				forgerInfo?.maxHeightPrevoted !== maxHeightPrevoted ||
+				forgerInfo?.maxHeightPreviouslyForged !== maxHeightPreviouslyForged
+			) {
+				throw new Error('Failed to enable forging due to contradicting forger info.');
+			}
+		}
+
+		previouslyForgedMap.set(forgerAddress, {
+			height,
+			maxHeightPrevoted,
+			maxHeightPreviouslyForged,
+		});
+		await setPreviouslyForgedMap(this._db, previouslyForgedMap);
+		this._logger.info(forgingInput, 'Updated forgerInfo');
```

Comment:
Isn't it a bit misleading if we log this even if `overwrite==False` and the forgerInfo actually stays the same?

In theory, line 235-241 could be in the else branch for the case `overwrite==True`, right?